src/main/java/com/atlassian/db/replica/api/AuroraPostgresLsnReplicaConsistency.java

```diff
@@ -0,0 +1,99 @@
+package com.atlassian.db.replica.api;
+
+import com.atlassian.db.replica.internal.MonotonicMemoryCache;
+import com.atlassian.db.replica.spi.Cache;
+import com.atlassian.db.replica.spi.ReplicaConsistency;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * It remembers last write LSN (log sequence number) and compares it with LSN for subsequent reads to detect inconsistencies.
+ * If it cannot remember the LSN of last write, pessimistically assumes it's going to be inconsistent.
+ *
+ * @see <a href="https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-global-database-monitoring.html">Monitoring Aurora PostgreSQL-based Aurora global databases</a>
+ * <p>
+ * It's an PostgreSQL-based Aurora database engine specific implementation.
+ */
+public final class AuroraPostgresLsnReplicaConsistency implements ReplicaConsistency {
+
+    private final Cache<Long> lastWrite;
+
+    public static final class Builder {
+        private Cache<Long> lastWrite = new MonotonicMemoryCache<>();
+
+        /**
+         * @param lastWrite remembers last write
+         */
+        public Builder cacheLastWrite(Cache<Long> lastWrite) {
+            this.lastWrite = lastWrite;
+            return this;
+        }
+
+        /**
+         * @return consistency assuming that LSN (log sequence number) is greater or equal to LSN for last write (if known)
+         */
+        public AuroraPostgresLsnReplicaConsistency build() {
+            return new AuroraPostgresLsnReplicaConsistency(lastWrite);
+        }
+    }
+
+    private AuroraPostgresLsnReplicaConsistency(Cache<Long> lastWrite) {
+        this.lastWrite = lastWrite;
+    }
+
+    @Override
+    public void write(Connection main) {
+        try {
+            lastWrite.put(queryMainDbLsn(main));
+        } catch (Exception e) {
```

Comment:
What do you suggest here? IMO we should add logger here but it's not yet added to library.

Reply:
Adding a logger (even as generic as slf4j) adds friction for the consumers, they'd have to:
* bridge their logging framework to ours (it's easy but not zero cost)
* synchronize on logging dependency version
* force English exception messages

Instead, we can rethrow and let the consumer choose their error handling (logging, retrying, etc.). The `reset` can fail too, so they'll have to do it anyway.

External retries would work on `write(Connection main)` level, rather than just the `queryMainDbLsn` part. If this granularity is insufficient, we can add `.put(Supplier<T> value)` method to the cache. But I'd wait with that until it's actually needed.

Reply:
> Adding a logger
Postgres JDBC uses `java.util.logging` to avoid adding dependencies.

Reply:
I'm not sure that we should rethrow exception and force api client to handle it. We know that in case of any error we should invalidate cache to make further reads inconsistent. What is missing is only part about "visibility" so maybe it's enough to log it using `java.util.logging`?

Reply:
JUL could work, nice find. But the `reset` itself is not so clear-cut.

The `reset` call can also fail, so it doesn't guarantee inconsistency passing anyway. There are also additional costs to this.

If cache is timing out, then `put` was already slow. Then we call `reset`, which will slow it down even more and probably time out too.

Also this is a check-then-act race condition, which can erase a legitimate new `lastWrite` from another request.

We shouldn't force this error handling. Let's ship the LSN part itself and leave error handling for another place. The consumer can choose to reset if they want in their own `ReplicaConsistency` layer (or in `Cache` if it gets `put` with `Supplier`).